armhf build issues with Atomic

During testing builds for Raspbian this error came up:

https://git.icinga.com/packaging/raspbian-icinga2/-/jobs/13153

```
[ 99%] Linking CXX executable ../Bin/None/boosttest-test-livestatus
cd /builds/packaging/raspbian-icinga2/build/icinga2/obj-arm-linux-gnueabihf/test && /usr/bin/cmake -E cmake_link_script CMakeFiles/boosttest-test-livestatus.dir/link.txt --verbose=1
/usr/lib/ccache/c++  -g -O2 -fdebug-prefix-map=/builds/packaging/raspbian-icinga2/build/icinga2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g -pthread -Winvalid-pch -flto -std=c++0x  -Wl,-z,relro -flto -rdynamic CMakeFiles/boosttest-test-livestatus.dir/test-runner.cpp.o CMakeFiles/boosttest-test-livestatus.dir/livestatus_unity.cpp.o ../third-party/mmatch/CMakeFiles/mmatch.dir/mmatch.c.o ../third-party/socketpair/CMakeFiles/socketpair.dir/socketpair.c.o ../lib/base/CMakeFiles/base.dir/application-version.cpp.o ../lib/base/CMakeFiles/base.dir/base_unity.cpp.o ../third-party/execvpe/CMakeFiles/execvpe.dir/execvpe.c.o ../lib/config/CMakeFiles/config.dir/config_lexer.cc.o ../lib/config/CMakeFiles/config.dir/config_parser.cc.o ../lib/config/CMakeFiles/config.dir/config_unity.cpp.o ../lib/remote/CMakeFiles/remote.dir/remote_unity.cpp.o ../lib/icinga/CMakeFiles/icinga.dir/icinga-itl.cpp.o ../lib/icinga/CMakeFiles/icinga.dir/icinga_unity.cpp.o ../lib/livestatus/CMakeFiles/livestatus.dir/livestatus_unity.cpp.o ../lib/methods/CMakeFiles/methods.dir/methods-itl.cpp.o ../lib/methods/CMakeFiles/methods.dir/methods_unity.cpp.o  -o ../Bin/None/boosttest-test-livestatus -ldl -lboost_context -lboost_coroutine -lboost_date_time -lboost_filesystem -lboost_thread -lpthread -lboost_system -lboost_program_options -lboost_regex -lboost_chrono -lboost_atomic -lssl -lcrypto -lsystemd -ledit -ltermcap -lboost_unit_test_framework 
/usr/bin/ld: /usr/bin/ld: DWARF error: invalid abstract instance DIE ref
/tmp/ccZmlvHf.ltrans0.ltrans.o: in function `icinga::ValidationError::ValidationError(icinga::ValidationError const&) [clone .constprop.894]':
<artificial>:(.text+0xe208): undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: <artificial>:(.text+0xe280): undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: /tmp/ccZmlvHf.ltrans0.ltrans.o: in function `__static_initialization_and_destruction_0(int, int) [clone .constprop.608]':
<artificial>:(.text.startup+0x2f2c): undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: /usr/bin/ld: DWARF error: invalid abstract instance DIE ref
/tmp/ccZmlvHf.ltrans1.ltrans.o: in function `std::__shared_count<(__gnu_cxx::_Lock_policy)1>::__shared_count<icinga::ConstEmbeddedNamespaceValue, std::allocator<icinga::ConstEmbeddedNamespaceValue>, boost::intrusive_ptr<icinga::Function>&>(icinga::ConstEmbeddedNamespaceValue*&, std::_Sp_alloc_shared_tag<std::allocator<icinga::ConstEmbeddedNamespaceValue> >, boost::intrusive_ptr<icinga::Function>&) [clone .constprop.607]':
<artificial>:(.text+0x6ed4): undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: /tmp/ccZmlvHf.ltrans1.ltrans.o: in function `void boost::call_once<EnsureFrameCleanupTimer()::{lambda()#1}>(boost::once_flag&, EnsureFrameCleanupTimer()::{lambda()#1}&&) [clone .isra.5046] [clone .constprop.24]':
<artificial>:(.text+0xd244): undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: /usr/bin/ld: DWARF error: invalid abstract instance DIE ref
/tmp/ccZmlvHf.ltrans7.ltrans.o:<artificial>:(.text+0x9de0): more undefined references to `__atomic_fetch_add_8' follow
/usr/bin/ld: /usr/bin/ld: DWARF error: invalid abstract instance DIE ref
/tmp/ccZmlvHf.ltrans18.ltrans.o: in function `(anonymous namespace)::io22::{lambda()#22}::operator()() const [clone .isra.4224] [clone .lto_priv.2270]':
<artificial>:(.text+0x999c): undefined reference to `__atomic_store_8'
/usr/bin/ld: <artificial>:(.text+0x9ab0): undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: <artificial>:(.text+0x9b78): undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: <artificial>:(.text+0x9c2c): undefined reference to `__atomic_fetch_add_8'
...
```

@dnsmichi suggested the following links:

* https://github.com/PowerDNS/pdns/issues/4171
* https://github.com/PurpleI2P/i2pd/issues/897

I'm currently preparing a Boost build for Raspbian stretch, and can then verify the result there.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -373,6 +373,54 @@ if(NOT MSVC)
   endif()
 endif()
 
+# Architecture specifics
+# - Log the target architecture
+# - ARM needs to link against atomic
+if(NOT MSVC)
+  # inspired by https://github.com/civetweb/civetweb/blob/master/cmake/DetermineTargetArchitecture.cmake
+  execute_process(
+    COMMAND ${CMAKE_C_COMPILER} -dumpmachine
+    RESULT_VARIABLE RESULT
+    OUTPUT_VARIABLE ARCH
+    ERROR_QUIET
+  )
+
+  if (RESULT)
+    message(FATAL_ERROR "Failed to detect target architecture: ${RESULT}")
+  endif()
+
+  string(REGEX MATCH "([^-]+).*" ARCH_MATCH ${ARCH})
+  if (NOT CMAKE_MATCH_1 OR NOT ARCH_MATCH)
+    message(FATAL_ERROR "Failed to match the target architecture: ${ARCH}")
+  endif()
+
+  set(ARCH ${CMAKE_MATCH_1})
+
+  message(STATUS "Target architecture - ${ARCH}")
+
+  # ARM settings
+  if("${ARCH}" STREQUAL "arm")
+    check_cxx_source_compiles( "include <atomic>; int main(){ std::atomic<uint_fast64_t> x; x.fetch_add(1); x.sub_add(1);  }" CXX_ATOMIC)
+
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -latomic")
+    set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -latomic")
+    set(CMAKE_STATIC_LINKER_FLAGS "${CMAKE_STATIC_LINKER_FLAGS} -latomic")
+  endif()
+
+else()
+  if("${MSVC_C_ARCHITECTURE_ID}" STREQUAL "X86")
+    set(ARCH "i686")
+  elseif("${MSVC_C_ARCHITECTURE_ID}" STREQUAL "x64")
+    set(ARCH "x86_64")
+  elseif("${MSVC_C_ARCHITECTURE_ID}" STREQUAL "ARM")
+    set(ARCH "arm")
+  else()
+    message(FATAL_ERROR "Failed to determine the MSVC target architecture: ${MSVC_C_ARCHITECTURE_ID}")
+  endif()
+
+  message(STATUS "Target architecture - ${ARCH}")
+endif()
+
 configure_file(config.h.cmake ${CMAKE_CURRENT_BINARY_DIR}/config.h ESCAPE_QUOTES)
 
 install(

